File: js/json/src/compression.ts
/**
 * Group-scoped DEFLATE compression for the JSON frame stream, using
 * {@link https://github.com/nodeca/pako | pako}'s streaming deflate/inflate.
 *
 * Within a group the frame payloads form a single raw DEFLATE
 * ([RFC 1951](https://www.rfc-editor.org/rfc/rfc1951.html)) stream, sync-flushed at each frame
 * boundary so every frame is self-delimited while later frames reuse the earlier ones as context
 * (a snapshot followed by deltas compresses far better than each frame alone). This matches the
 * Rust `moq-json` producer, so the two interoperate on the wire.
 *
 * A sync flush always ends in the fixed 4-byte marker `00 00 ff ff`. {@link Encoder.frame} drops
 * it and {@link Decoder.frame} re-appends it, saving 4 bytes per frame, the same trick
 * [RFC 7692](https://www.rfc-editor.org/rfc/rfc7692.html#section-7.2.1) (permessage-deflate) uses.
 * moq-net frames each slice, so there's no length prefix; {@link Decoder.frame} instead caps the
 * inflated output as it is produced.
 *
 * pako is synchronous, so the whole codec is synchronous; it is a normal dependency.
 *
 * @module
 */

import * as pako from "pako";

// Maximum decompressed size of a single frame. A malicious publisher could otherwise send a tiny
// slice that inflates hugely, so {@link Decoder} stops retaining output past this and rejects the
// frame. Mirrors the Rust `MAX_DECOMPRESSED_FRAME`.
const MAX_DECOMPRESSED_FRAME = 64 * 1024 * 1024;

// The trailing bytes of a DEFLATE sync flush, stripped on the wire and re-appended to decode.
const SYNC_FLUSH_TAIL = new Uint8Array([0x00, 0x00, 0xff, 0xff]);

// Concatenate chunks into one tight buffer (a single chunk passes through untouched). Safe only for
// output that is consumed before the next push, since a single chunk is pako's own reused buffer.
function concat(chunks: Uint8Array[], total: number): Uint8Array {
	if (chunks.length === 1) return chunks[0];
	const out = new Uint8Array(total);
	let offset = 0;
	for (const chunk of chunks) {
		out.set(chunk, offset);
		offset += chunk.length;
	}
	return out;
}

/**
 * Encodes a group's frame payloads into one shared DEFLATE stream, one self-delimited slice per
 * frame. Hold one per group; create a new one at each group boundary.
 *
 * @public
 */
export class Encoder {
	#deflate = new pako.Deflate({ raw: true });
	#chunks: Uint8Array[] = [];
	#total = 0;

	/** Start a fresh per-group encoder with a cold window. */
	constructor() {
		this.#deflate.onData = (chunk) => {
			const bytes = chunk as Uint8Array;
			this.#chunks.push(bytes);
			this.#total += bytes.length;
		};
	}

	/**
	 * Compress the next frame's `payload`, returning its slice of the group stream: the DEFLATE bytes
	 * minus the fixed sync-flush marker. Empty in yields empty out. Slices must be produced in frame
	 * order.
	 */
	frame(payload: Uint8Array): Uint8Array {
		if (payload.length === 0) return payload;
		this.#chunks = [];
		this.#total = 0;
		this.#deflate.push(payload, pako.constants.Z_SYNC_FLUSH);

		// Copy into one tight owned buffer, dropping the trailing sync-flush marker. We can't return
		// pako's chunk views: `writeFrame` retains the reference and pako backs each chunk with a
		// ~16 KB buffer, so a view would pin far more memory than the frame.
		const out = new Uint8Array(this.#total - SYNC_FLUSH_TAIL.length);
		let offset = 0;
		for (const chunk of this.#chunks) {
			if (offset >= out.length) break;
			const take = Math.min(chunk.length, out.length - offset);
			out.set(chunk.subarray(0, take), offset);
			offset += take;
		}
		return out;
	}
}

/**
 * Decodes a group's frame slices back into the original payloads. Hold one per group; feed slices
 * in frame order (each frame builds on the earlier ones).
 *
 * @public
 */
export class Decoder {
	#inflate = new pako.Inflate({ raw: true });
	#chunks: Uint8Array[] = [];
	#total = 0;
	#tooLarge = false;

	/** Start a fresh per-group decoder with a cold window. */
	constructor() {
		this.#inflate.onData = (chunk) => {
			const bytes = chunk as Uint8Array;
			this.#total += bytes.length;
			// Bound the inflated output as it is produced; a tiny slice can expand enormously. Stop
			// retaining past the cap, then reject once the push returns.
			if (this.#total > MAX_DECOMPRESSED_FRAME) {
				this.#tooLarge = true;
				return;
			}
			this.#chunks.push(bytes);
		};
	}

	/**
	 * Decompress the next frame's `slice` back into its payload. Empty in yields empty out. Throws if
	 * the input is malformed or inflates past the per-frame size limit.
	 */
	frame(slice: Uint8Array): Uint8Array {
		if (slice.length === 0) return slice;

		this.#chunks = [];
		this.#total = 0;
		this.#tooLarge = false;

		// Feed the slice then the re-appended sync-flush marker as two pushes, so no combined buffer is
		// allocated. The marker delimits the frame and flushes its last bytes out of the inflate buffer.
		this.#inflate.push(slice, false);
		this.#inflate.push(SYNC_FLUSH_TAIL, pako.constants.Z_SYNC_FLUSH);
		if (this.#inflate.err) throw new Error(`decompression failed: ${this.#inflate.msg}`);
		if (this.#tooLarge) throw new Error(`decompressed frame exceeded ${MAX_DECOMPRESSED_FRAME} bytes`);

		return concat(this.#chunks, this.#total);
	}
}

File: rs/moq-json/src/compression.rs
//! Group-scoped DEFLATE compression for the JSON frame stream.
//!
//! Within a group the frame payloads form a single raw DEFLATE ([RFC 1951]) stream, sync-flushed
//! at each frame boundary so every frame carries its own self-delimited slice while later frames
//! reuse the earlier ones as context (a snapshot followed by deltas compresses far better than
//! each frame alone). The [`Encoder`]/[`Decoder`] hold that per-group state; both are recreated at
//! every group boundary.
//!
//! This is plain raw DEFLATE with a `Z_SYNC_FLUSH` after each frame, so a browser (`@moq/json`)
//! peer interoperates on the wire using the same primitive (zlib's sync flush). moq-net already
//! frames each slice, so there's no length prefix. A small slice can still inflate to far more than
//! its own size, so [`Decoder::frame`] bounds each frame's output at [`MAX_DECOMPRESSED_FRAME`].
//!
//! A sync flush always ends in the 4-byte empty-block marker `00 00 ff ff`. That marker is fixed,
//! so [`Encoder::frame`] drops it from each slice and [`Decoder::frame`] re-appends it before
//! inflating, saving 4 bytes per frame. This is the same trick [RFC 7692] (permessage-deflate)
//! uses for WebSocket messages.
//!
//! [RFC 1951]: https://www.rfc-editor.org/rfc/rfc1951.html
//! [RFC 7692]: https://www.rfc-editor.org/rfc/rfc7692.html#section-7.2.1

use bytes::Bytes;
use flate2::{Compress, Decompress, FlushCompress, FlushDecompress, Status};

use crate::{Error, Result};

/// DEFLATE level for the frame stream: zlib's own default, a good size/speed balance for the small,
/// repetitive payloads this targets.
const LEVEL: u32 = 6;

/// The trailing bytes of a DEFLATE sync flush, stripped on the wire and re-appended to decode.
const SYNC_FLUSH_TAIL: [u8; 4] = [0x00, 0x00, 0xff, 0xff];

/// Maximum decompressed size of a single frame.
///
/// A malicious publisher could otherwise send a tiny slice that inflates hugely, so
/// [`Decoder::frame`] stops and returns [`Error::TooLarge`] rather than allocating without limit.
const MAX_DECOMPRESSED_FRAME: u64 = 64 * 1024 * 1024;

/// Scratch buffer size for the streaming (de)compress loops.
const CHUNK: usize = 8 * 1024;

/// Encodes a group's frame payloads into one shared DEFLATE stream, one self-delimited slice per
/// frame. Hold one per group; the stream is recreated at each group boundary.
pub(crate) struct Encoder(Compress);

impl Encoder {
	/// Start a fresh per-group encoder with a cold window.
	pub(crate) fn new() -> Self {
		// `false`: raw DEFLATE, no zlib header/trailer, matching `deflate-raw` on the browser side.
		Self(Compress::new(flate2::Compression::new(LEVEL), false))
	}

	/// Compress the next frame's `payload`, returning its slice of the group stream: the DEFLATE
	/// bytes minus the fixed sync-flush marker. Empty in yields empty out. Later frames reuse earlier
	/// ones as context, so slices must be produced (and later decoded) in frame order.
	pub(crate) fn frame(&mut self, payload: &[u8]) -> Bytes {
		if payload.is_empty() {
			return Bytes::new();
		}

		let mut out = Vec::with_capacity(payload.len() / 2 + 16);
		let mut tmp = [0u8; CHUNK];
		let mut input = payload;

		// Drive the stream with a sync flush so this frame's slice is self-delimited (byte-aligned,
		// window retained). The classic zlib loop: keep going while the output buffer fills up.
		loop {
			let before_in = self.0.total_in();
			let before_out = self.0.total_out();
			self.0.compress(input, &mut tmp, FlushCompress::Sync).expect("deflate");
			let consumed = (self.0.total_in() - before_in) as usize;
			let produced = (self.0.total_out() - before_out) as usize;
			out.extend_from_slice(&tmp[..produced]);
			input = &input[consumed..];
			if produced < tmp.len() {
				break;
			}
		}

		// Drop the fixed sync-flush marker; the decoder re-appends it (see the module docs).
		debug_assert!(
			out.ends_with(&SYNC_FLUSH_TAIL),
			"a sync flush must end in the deflate marker"
		);
		out.truncate(out.len() - SYNC_FLUSH_TAIL.len());
		Bytes::from(out)
	}
}

/// Decodes a group's frame slices back into the original payloads. Hold one per group; feed slices
/// in frame order (each frame builds on the earlier ones).
pub(crate) struct Decoder(Decompress);

impl Decoder {
	/// Start a fresh per-group decoder with a cold window.
	pub(crate) fn new() -> Self {
		// `false`: raw DEFLATE, matching the encoder.
		Self(Decompress::new(false))
	}

	/// Decompress the next frame's `slice` back into its payload.
	///
	/// An empty slice yields an empty payload. Returns [`Error::TooLarge`] if the frame inflates past
	/// the per-frame bound (checked as output is produced, not from any declared size), and
	/// [`Error::Decompress`] on malformed input.
	pub(crate) fn frame(&mut self, slice: &[u8]) -> Result<Bytes> {
		if slice.is_empty() {
			return Ok(Bytes::new());
		}

		let mut out = Vec::new();
		let mut tmp = [0u8; CHUNK];

		// Feed the wire slice followed by the re-appended sync-flush marker, which delimits the frame
		// and flushes its last bytes out of the inflate buffer.
		for segment in [slice, &SYNC_FLUSH_TAIL] {
			let mut input = segment;
			loop {
				let before_in = self.0.total_in();
				let before_out = self.0.total_out();
				let status = self
					.0
					.decompress(input, &mut tmp, FlushDecompress::Sync)
					.map_err(|_| Error::Decompress)?;
				let consumed = (self.0.total_in() - before_in) as usize;
				let produced = (self.0.total_out() - before_out) as usize;
				// Bound the inflated output as it is produced; a tiny slice can expand enormously.
				if out.len() as u64 + produced as u64 > MAX_DECOMPRESSED_FRAME {
					return Err(Error::TooLarge(MAX_DECOMPRESSED_FRAME));
				}
				out.extend_from_slice(&tmp[..produced]);
				input = &input[consumed..];

				// Move to the next segment once this one is drained and the buffer wasn't saturated. The
				// no-progress guard avoids spinning when the marker needs no further output.
				if matches!(status, Status::StreamEnd) || (input.is_empty() && produced < tmp.len()) {
					break;
				}
				if consumed == 0 && produced == 0 {
					break;
				}
			}
		}

		Ok(Bytes::from(out))
	}
}

#[cfg(test)]
mod test {
	use super::*;

	/// Round-trip a sequence of frames through a group encoder/decoder pair.
	fn roundtrip(frames: &[&[u8]]) -> Vec<Vec<u8>> {
		let mut enc = Encoder::new();
		let slices: Vec<Bytes> = frames.iter().map(|f| enc.frame(f)).collect();

		let mut dec = Decoder::new();
		slices.iter().map(|s| dec.frame(s).unwrap().to_vec()).collect()
	}

	#[test]
	fn group_roundtrip() {
		let frames: &[&[u8]] = &[b"the quick brown fox", b"the quick brown dog", b"the lazy fox"];
		let got = roundtrip(frames);
		for (a, b) in frames.iter().zip(&got) {
			assert_eq!(*a, b.as_slice());
		}
	}

	#[test]
	fn empty_frames_roundtrip() {
		assert!(Encoder::new().frame(b"").is_empty());
		assert!(Decoder::new().frame(b"").unwrap().is_empty());
	}

	#[test]
	fn cross_frame_context_shrinks() {
		// A later frame identical to an earlier one compresses to far fewer bytes once the window
		// holds the earlier copy: this is the whole point of a shared per-group stream.
		let payload = b"Media over QUIC delivers real-time latency at massive scale.".repeat(6);
		let mut enc = Encoder::new();
		let first = enc.frame(&payload);
		let second = enc.frame(&payload);
		assert!(
			second.len() < first.len(),
			"repeat frame {} should be smaller than first {}",
			second.len(),
			first.len()
		);
	}

	#[test]
	fn frame_larger_than_chunk_roundtrips() {
		// High-entropy data barely compresses, so its slice exceeds the streaming `CHUNK` scratch
		// buffer and the (de)compress loops must iterate. Verify it still round-trips byte for byte.
		let mut state: u64 = 0x9E37_79B9_7F4A_7C15;
		let payload: Vec<u8> = (0..64 * 1024)
			.map(|_| {
				state ^= state << 13;
				state ^= state >> 7;
				state ^= state << 17;
				(state >> 56) as u8
			})
			.collect();

		let mut enc = Encoder::new();
		let slice = enc.frame(&payload);
		assert!(slice.len() > CHUNK, "slice {} should exceed CHUNK {CHUNK}", slice.len());

		let mut dec = Decoder::new();
		assert_eq!(dec.frame(&slice).unwrap(), Bytes::from(payload));
	}

	#[test]
	fn decompress_rejects_garbage() {
		let mut dec = Decoder::new();
		assert!(matches!(
			dec.frame(b"not a deflate stream at all"),
			Err(Error::Decompress)
		));
	}
}
